### DISCUSSION TITLE
None is not allowed value for Required Optional objects

### DISCUSSION THREAD

**Discussion by ghost** [pydantic/pydantic#4872] [Question] [answered]
Hi,
my problem is a bit different from the one previously asked by other users.

My JSON file got one lines like `"itemNumber": null,` but in the same file there are several other lines with values i.e. `"itemNumber": "1740", `

My python script should check if the file is valid

    import json
    from pydantic import BaseModel, Field
    import pydantic
    from typing import Any, List, Optional
    
    class MeasuringDevice(BaseModel):
        itemNumber: Optional[str] = Field(..., nullable=True)

    class Model(BaseModel):
        measuringDevices: List[MeasuringDevice]
    
    json_item = json.loads(open(entry, 'r').read())
    try:
        Model(**json_item)
    except pydantic.ValidationError as exc:
        print(f"ERROR: Invalid schema: {exc}")
        return False

So I used 
`itemNumber: Optional[str]` in my model, which didn't work, and I also tried<br/>
`itemNumber: Optional[str] = ...` and<br/> 
`itemNumber: Optional[str] = Field(..., nullable=True)` and<br/> 
`itemNumber: Any` but all the time I receive<br/>
`none is not an allowed value (type=type_error.none.not_allowed)` as long as the `"itemNumber": null,` exists in the file.<br/><br/> 
While other items with null values (Consistent "null" in all appearances) did not raise this error.

How can I prevent this from happen? I got 13 Option fields in my model and only 5 test files yet.

**Comment by uriyyo** [comment, marked as answer]
Hi @argv1,

This issue will be fixed in `pydantic v2` - https://docs.pydantic.dev/blog/pydantic-v2/#required-vs-nullable-cleanup

For this moment you can try to use `prapare_field` config hook to make `Optional` fields required.
```python
import json
from typing import List, Optional

from pydantic import BaseModel, BaseConfig, ValidationError
from pydantic.fields import ModelField
from pydantic.typing import is_union, get_args, get_origin


class MeasuringDevice(BaseModel):
    itemNumber: Optional[str]

    class Config(BaseConfig):
        @classmethod
        def prepare_field(cls, field: ModelField) -> None:
            # check if field is Optional
            if is_union(get_origin(field.outer_type_)) and type(None) in get_args(field.outer_type_):
                field.required = True


class Model(BaseModel):
    measuringDevices: List[MeasuringDevice]


json_item = json.loads(
    """
    {
        "measuringDevices": [
            {
                "itemNumber": null
            },
            {
                "itemNumber": "123"
            }
        ]
    }
    """
)
try:
    Model(**json_item)
except ValidationError as exc:
    print(f"ERROR: Invalid schema: {exc}")
```

**Reply by KayvanShah1** [reply]
This doesn't work when using this model as a Response Model

**Reply by matheushent** [reply]
forgive me but, what's the sense on turning an intentional optional field as required??

**Reply by dmontagu** [reply]
In python, what annotation of `x: Optional[T]` _actually_ means is that `x` can be of type `T` or be the value `None`. This is how the python `typing` module works — `Optional[T]` has the _exact same_ meaning as `Union[T, None]`. It doesn't mean you don't have to specify the value — just that you are allowed to pass `None`.

This is how both standard library dataclasses and standard python functions work:
```python
def f(x: Optional[int]):
    ...
```
In this case, you would need to pass an argument for `x`.

In all cases, however, you can actually make the argument optional by setting a default (generally None for optional).

In v1 of pydantic, we did this automatically, but in v2 we have moved to this behavior, which I think is generally more consistent with how other standard module libraries and the language itself have been designed.
